src/pages/SearchPage.tsx

```diff
@@ -0,0 +1,179 @@
+        console.log(openDetails);
+      }
+    }
+  }
+  async function getSearchResults(
+    searchStr: string,
+    limit: number,
+    currentPage: number
+  ) {
+    setIsSearchLoading(true);
+    try {
+      const responseJSON = await APISearch(searchStr, limit, currentPage);
+      const resultsJSON = responseJSON.docs;
+      const totalPagesData = responseJSON.pages;
+      setTotalPages(totalPagesData);
+
+      const results = resultsJSON.map((item: SearchResultType) => ({
+        name: item.name,
+        height: item.height,
+        race: item.race,
+        birth: item.birth,
+        spouse: item.spouse,
+        _id: item._id,
+      }));
+      setSearchResults(results);
+    } catch (error) {
+      console.log(error);
+      setHasError(true);
+    } finally {
+      setIsSearchLoading(false);
+    }
+  }
+
+  async function setCharacterDetails() {
+    setAreDetailsLoading(true);
+    try {
+      const responseJSON = await APICharacterID(currentItemID);
+      const resultsJSON = responseJSON.docs;
+      const result = currentItemIDInitial;
+      Object.assign(result, resultsJSON[0]);
+      setCurrentItem(result);
+    } catch (error) {
+      console.log(error);
+      setHasError(true);
+    } finally {
+      setAreDetailsLoading(false);
+    }
+  }
+
+  async function sendRequest() {
+    setCurrentPage(1);
+    getSearchResults(query, limit, currentPage);
+    navigate('/results/1');
+    localStorage.setItem('searchQuery', query.trim());
+  }
+
+  function showDetails(id: string) {
+    setCurrentItemID(id);
+    setOpenDetails(true);
+  }
+
+  function makeError() {
+    setHasError(true);
+    throw new Error('New Error');
+  }
+
+  return (
+    <SearchResultsContext.Provider value={{ searchResults, currentItem }}>
+      <LoadingContext.Provider
+        value={{ areDetailsLoading, setAreDetailsLoading }}
+      >
+        <SearchContext.Provider value={{ query, setQuery }}>
+          <div onClick={handleClick}>
+            <SearchBar
+              className="search-input"
+              type="text"
+              query={query}
+              placeholder={placeholderText}
+              sendRequest={sendRequest}
+              makeError={makeError}
+            ></SearchBar>
+            <ItemsSelect getValue={changeItemsLimit}></ItemsSelect>
+            <SearchResults
+              isSearchLoading={isSearchLoading}
+              showDetails={showDetails}
+            />
+            {isSearchLoading ? null : (
+              <Pagination
+                currentPage={currentPage}
+                changePage={changePage}
+                totalPages={totalPages}
+              />
+            )}
+            {openDetails ? (
+              <DetailsSection closeDetails={() => setOpenDetails(false)} />
+            ) : null}
+          </div>
+        </SearchContext.Provider>
+      </LoadingContext.Provider>
+    </SearchResultsContext.Provider>
+  );
+}
+
+export default SearchPage;
```

Comment:
It would be great to refactor this component and move part of logic into helpers or hooks